[plugin] Implement SCM Repository selected event

vinokurig opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->
Implement SCM Repository selected event: https://github.com/microsoft/vscode/blob/a50cf110d96fbe50fc460a6de14ade5b6b060dd4/src/vs/vscode.proposed.d.ts#L673-L684
Removed duplicated `SourceControlResourceDecorations` interface.
#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->
1. Open a workspace with more than one Git repositories.
2. Clone to the workspace folder and compile the test plugin: https://github.com/vinokurig/Test.git
3. Start hosted instance with the plugin.
4. Select the `/` SCM repository (this is a view of a dummy SCM repository):
![screenshot-localhost-3030-2019 09 06-17-34-24](https://user-images.githubusercontent.com/7668752/64436695-892fe000-d0cd-11e9-8f13-ccd7492bcbb7.png)

See the notification that is fired inside the event handler.
Select other repository again.
See the notification that the repository is **NOT** selected.
#### Review checklist

- [ ] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

akosyakov commented on packages/plugin-ext/src/main/browser/scm-main.ts:
```diff
@@ -37,12 +37,29 @@ export class ScmMainImpl implements ScmMain {
+                    }
+                });
+            }
+            if (!externalRepository && this.selectedSourceControlHandle !== undefined) {
```
Confused about `externalRepository` - does it mean something selected? If so should not we unselect previous selection anyway, i.e.
```
const sourceControlHandle = this.getSourceControlHandle(repository);
if (sourceControlHandle !== this.selectedSourceControlHandle) {
   this.proxy.$setSourceControlSelection(this.selectedSourceControlHandle, sourceControlHandle); // old, new, both of `number | undefined` type
   this.selectedSourceControlHandle = sourceControlHandle;
}

vinokurig replied:
`externalRepository` means that the selected repository is registered from the plugin, so it has a handle id (the repository exists in the `scmRepositoryMap`). The repository may not be present in the `scmRepositoryMap` if it is registered from internal extension e.g. built in Git extension. In this  case we need to deselect previously selected repository that has a handle id.

akosyakov replied:
I cannot understand whether `this.proxy.$setSourceControlSelection(this.selectedSourceControlHandle, false);` will be called in the case if `this.proxy.$setSourceControlSelection(key, true);` is called above. From code it looks like it won't because `!externalRepository`. So it means that the previous source control won't be unselected.

akosyakov replied:
I also don't get why internal/external is important. I thought the logic should be unselect last handle if it is not the same as current. If repository is internal then handle will be undefined and we select nothing, but still unselect the previous handle.

vinokurig replied:
Reworked the logic

akosyakov replied:
the logic is straight to understand now, thank you!

akosyakov commented:
@vinokurig I could not figure out how I can test that a source control is properly unselected with the  proposed extension. We need an extensions which has several source controls and show notifications when one get unselected or selected. `@theia/git` extension should be removed during testing.

vinokurig commented:
@akosyakov I think I've understood your point, I've reworked the logic and now you can test this against the plugin provided above that I've changed as well. To apply the plugin you need to disable built in Git extension and paste any folder paths as a repository url: https://github.com/vinokurig/Test/blob/2f48915bab1487d5998ceed06c68d60c898f5ffc/src/scm-plugin-backend.ts#L9-L11

akosyakov commented:
@vinokurig it looks better now! but i don't see initial notification and after i switch to another repo there is no notification that an initial repository is unselected.

akosyakov commented on packages/plugin-ext/src/main/browser/scm-main.ts:
```diff
@@ -37,12 +37,30 @@ export class ScmMainImpl implements ScmMain {
         this.scmRepositoryMap = new Map();
         this.labelProvider = container.get(LabelProvider);
+        this.scmService.onDidChangeSelectedRepository(repository => {
+            const sourceControlHandle = repository ? this.getSourceControlHandle(repository) : undefined;
```
maybe we are better to extract a method `updateSelectedRepository` which calls `this.scmService.selectedRepository`

When we call it initially and from event handler.

vinokurig replied:
done

vinokurig commented:
@akosyakov
> @vinokurig it looks better now! but i don't see initial notification and after i switch to another repo there is no notification that an initial repository is unselected.

I've fixed that by extracting the logic to a method as you recommended.

akosyakov commented on packages/plugin-ext/src/main/browser/scm-main.ts:
```diff
@@ -37,12 +37,32 @@ export class ScmMainImpl implements ScmMain {
+        this.lastSelectedSourceControlHandle = sourceControlHandle;
+    }
+
+    getSourceControlHandle(repository: ScmRepository): number | undefined {
```
`protected`

akosyakov commented:
@vinokurig please rebase and if ci is green please merge, don't forget to address the outstanding comments
